src/main/java/org/jenkinsci/plugins/jvctb/config/CredentialsHelper.java

```diff
@@ -1,96 +1,110 @@
 import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
 import com.cloudbees.plugins.credentials.domains.DomainRequirement;
+import com.cloudbees.plugins.credentials.domains.URIRequirementBuilder;
 import com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl;
 import com.google.common.base.Optional;
 
-import hudson.model.ItemGroup;
+import hudson.model.Item;
+import hudson.model.Queue;
+import hudson.model.queue.Tasks;
 import hudson.security.ACL;
+import hudson.util.FormValidation;
 import hudson.util.ListBoxModel;
 import hudson.util.Secret;
 import jenkins.model.Jenkins;
 
 public class CredentialsHelper {
-  public static ListBoxModel doFillUsernamePasswordCredentialsIdItems() {
-    final List<StandardUsernamePasswordCredentials> credentials = getAllCredentials();
-    final AbstractIdCredentialsListBoxModel<
-            StandardUsernameListBoxModel, StandardUsernameCredentials>
-        listBoxModel = new StandardUsernameListBoxModel().includeEmptyValue();
-    for (final StandardUsernamePasswordCredentials credential : credentials) {
-      listBoxModel.with(credential);
+  public static ListBoxModel doFillCredentialsIdItems(
+      Item item, String credentialsId, String uri) {
+    StandardListBoxModel result = new StandardListBoxModel();
+    if (item == null) {
+      if (!Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER)) {
+        return result.includeCurrentValue(credentialsId);
+      }
+    } else {
+      if (!item.hasPermission(Item.EXTENDED_READ)
+          && !item.hasPermission(CredentialsProvider.USE_ITEM)) {
+        return result.includeCurrentValue(credentialsId);
+      }
     }
-    return listBoxModel;
+    return result //
+        .includeEmptyValue() //
+        .includeMatchingAs(
+            item instanceof Queue.Task ? Tasks.getAuthenticationOf((Queue.Task) item) : ACL.SYSTEM,
+            item,
+            StandardCredentials.class,
+            URIRequirementBuilder.fromUri(uri).build(),
+            CredentialsMatchers.anyOf(
+                CredentialsMatchers.instanceOf(StandardUsernamePasswordCredentials.class),
+                CredentialsMatchers.instanceOf(StringCredentials.class)))
+        .includeCurrentValue(credentialsId);
   }
 
-  public static ListBoxModel doFillPersonalAccessTokenIdItems() {
-    final List<StringCredentials> credentials = getAllCredentials(StringCredentials.class);
-    final ListBoxModel listBoxModel =
-        new StandardListBoxModel() //
-            .includeEmptyValue() //
-            .withAll(credentials);
-    return listBoxModel;
-  }
-
-  public static Optional<StandardUsernamePasswordCredentials> findCredentials(
-      final String usernamePasswordCredentialsId) {
-    if (isNullOrEmpty(usernamePasswordCredentialsId)) {
-      return absent();
+  public static FormValidation doCheckFillCredentialsId(
+      Item item, String credentialsId, String uri) {
+    if (item == null) {
```

Comment:
What scenarios does these if-statements handle? What do they mean? Perhaps the booleans can be assigned to variables with some helpful naming.

Reply:
Fully explained in the docs: https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc

>We protect against fully populating the drop-down list for users that have no ability to actually make a selection. This is also useful in preventing unwanted requests being made to an external credentials store.

Reply:
Ok! Perhaps that link can be added to the code as a comment?